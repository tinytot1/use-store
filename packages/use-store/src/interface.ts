import React from "react"

// eslint-disable-next-line @typescript-eslint/no-invalid-void-type
export type ActionHandler = <T = any>(value: T, state: State) => void | Promise<void>;
export interface Actions {
  [key: string]: ActionHandler;
}

export type State = Record<string, any>;

export interface Model {
  name: string;
  state: State;
  actions?: Actions;
}

export interface ProviderProps {
  models: Model[];
  children?: React.ReactNode;
}

export interface SimpleProviderProps {
  model: Model;
  children?: React.ReactNode;
}

export type StoreDispatch = <T = any>(actionKey: string, value: T) => void;

export interface Status {
  pending: boolean;
  error: Error;
}