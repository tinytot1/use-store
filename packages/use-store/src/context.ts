import React, { useContext } from 'react';
import { Store } from './store';

const ContextMap = new Map<string, React.Context<Store>>();
const StoreMap = new Map<string, Store>();

export function getContext(name: string) {
  return ContextMap.get(name) || null;
}

export function createContext(name: string, store: Store): React.Context<Store> {
  const Context = React.createContext(store);
  // Ref https://zh-hans.reactjs.org/docs/context.html#contextdisplayname
  Context.displayName = name;
  ContextMap.set(name, Context);
  StoreMap.set(name, store);
  return Context;
}

export function deleteContext(name: string): void {
  ContextMap.delete(name);
  StoreMap.delete(name);
}

export function getStore(name: string) {
  return StoreMap.get(name);
}
