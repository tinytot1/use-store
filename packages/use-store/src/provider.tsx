import React, { useEffect, useMemo, useRef, useContext, useDebugValue, useReducer } from "react"

import { ProviderProps, SimpleProviderProps, State, StoreDispatch, Model, Status } from './interface';
import { createStore } from './store';
import { getContext, createContext, deleteContext } from './context';
import invariant from 'invariant';
import { looseEqual, isPlainObject, isFunction, isString } from './utils';
import { ACTION_STATUS_NAME as ASN, actionStatusModel, DEFAULT_STATUS } from './statusModel';

export function checkModels(models: Model[]) {

  invariant(Array.isArray(models), 'models should be array!');
  invariant(models.length >= 1, 'should provide at last one model!');

  for (const model of models) {
    checkModel(model);
  }
}

function checkModel(model: Model) {
  invariant(isPlainObject(model), 'model should be plain object!');

  const { name, state, actions = {} /* , middlewares */ } = model;

  invariant(name, 'model name is required!');
  invariant(typeof name === 'string', 'model name should be string!');

  invariant(isPlainObject(state), 'model state should be plain object!');

  invariant(isPlainObject(actions), `model actions should be plain object!`);
  for (const [action, handler] of Object.entries(actions)) {
    invariant(isFunction(handler), `model[${name}].actions[${action}] should be function!`);
  }

  // if (middlewares) checkMiddlewares(middlewares);
}

const SimpleProvider = React.memo<SimpleProviderProps>(({ model, children }) => {
  const { name } = model
  let Context = getContext(name);
  const store = useMemo(() => createStore(model), [model]);
  if (!Context) {
    Context = createContext(name, store);
  }

  useEffect(() => {
    // cleanup context
    return () => {
      deleteContext(name);
    };
  }, [name]);

  return <Context.Provider value={store}>{children}</Context.Provider>
})

// Provider HOC
const Provider = React.memo<ProviderProps>(({ models, children }) => {
  checkModels(models);
  models.push(actionStatusModel);
  let providers: React.ReactElement;
  // 消费多个 Context Ref https://zh-hans.reactjs.org/docs/context.html#consuming-multiple-contexts
  models.forEach(model => {
    providers = <SimpleProvider model={model}> {providers || children}</SimpleProvider>
  })
  return providers
})

export const createSharedStateContext = () => {
  return { Provider }
}

export function useStore<T = State>(name: string, selector: ((state: State) => T) = (s: State) => (s as T), isEqual = looseEqual): [T, StoreDispatch] {
  const Context = getContext(name);
  invariant(
    Context,
    `store with name[${name}] has not created, please ensure the component is wrapped in a <Provider>`,
  );

  // Ref https://zh-hans.reactjs.org/docs/hooks-reference.html#usedebugvalue
  useDebugValue(name)

  const store = useContext(Context);
  const prevRef = useRef<T>(selector(store.state));
  const value = prevRef.current;

  // Ref https://zh-hans.reactjs.org/docs/hooks-faq.html#is-there-something-like-forceupdate
  const forceUpdate = useReducer(x => !x, false)[1];

  useEffect(() => {
    const checkUpdate = (state: State) => {
      const newValue = selector(state)

      if (isEqual(value, newValue)) {
        return
      }
      prevRef.current = newValue;

      forceUpdate()
    }
    return store.subscribe(checkUpdate)
  }, [store, value, selector, forceUpdate, isEqual]);

  return [value, store.dispatch.bind(store)]
}

export function useStatus(actionWithName: string): [Status] {
  invariant(
    actionWithName && isString(actionWithName),
    'You must pass [name/action] to useStatus()',
  );
  const Context = getContext(ASN);

  // Ref https://zh-hans.reactjs.org/docs/hooks-reference.html#usedebugvalue
  useDebugValue(name)

  const store = useContext(Context);
  // init async action status when call `useStatus`
  const prevState = store.state;
  if (!prevState[actionWithName]) {
    prevState[actionWithName] = DEFAULT_STATUS;
  }
  
  const prevStatusRef = useRef<Status>(store.state[actionWithName]);
  const status: Status = prevStatusRef.current;

  // Ref https://zh-hans.reactjs.org/docs/hooks-faq.html#is-there-something-like-forceupdate
  const forceUpdate = useReducer(x => !x, false)[1];

  useEffect(() => {
    const checkStatus = (state: State) => {
      const newStatus = state[actionWithName];

      if (looseEqual(newStatus, status)) {
        return;
      }
      prevStatusRef.current = newStatus;

      forceUpdate();
    }
    return store.subscribe(checkStatus)
  }, [store, status, forceUpdate, actionWithName]);

  return [status]
}