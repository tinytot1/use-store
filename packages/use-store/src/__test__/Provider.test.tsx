import React, { useState, useContext } from "react";
import { mount } from "enzyme";
import { createSharedStateContext, useStore, useStatus } from "../provider";
import { act } from "react-dom/test-utils";
import { State } from '../interface';
import { getContext } from '../context';

describe("basic usage", () => {

  it('use-store should work', async () => {
    const models = [{
      name: 'modelA',
      state: { a: 0, b: 0 }
    }];

    const App = () => {
      const [{ a }] = useStore('modelA');
      return <button>a:{a}</button>;
    };
    const Context = createSharedStateContext()
    const Test = () => <Context.Provider models={models}><App /></Context.Provider>;
    const wrapper = mount(<Test />);
    expect(wrapper).toMatchSnapshot()
  })

  it('use-store should work when state changed', async () => {
    const models = [{
      name: 'modelA',
      state: { a: 0, b: 0 },
      actions: {
        addA(value: any, state: State) {
          state.a = value;
        }
      },
    }];

    const App = () => {
      const [a, dispatch] = useStore('modelA', s => s.a);
      const handleClick = (e: any) => {
        dispatch("addA", e.target.value)
      }
      return <button onClick={handleClick}>a:{a}</button>;
    };
    const AppB = () => {
      const [a] = useStore('modelA', s => s.a);
      return <div >a:{a}</div>;
    };
    const Context = createSharedStateContext()
    const Test = () => <Context.Provider models={models}><App /><AppB /></Context.Provider>;
    const wrapper = mount(<Test />);
    const buttonNode = wrapper.find("button").at(0);
    const appBNode = wrapper.find("AppB").find("div").at(0);
    expect(buttonNode.text()).toEqual('a:0');
    expect(appBNode.text()).toEqual('a:0');
    await act(async () => {
      buttonNode.simulate("click", { target: { value: 1 } })
      await (global as any).flushPromise(30);
    });
    wrapper.update();
    jest.runAllImmediates()
    expect(buttonNode.text()).toEqual('a:1');
    expect(appBNode.text()).toEqual('a:1');

    await act(async () => {
      buttonNode.simulate("click", { target: { value: 1 } })
      await (global as any).flushPromise(30);
    });
    wrapper.update();
    jest.runAllImmediates()
    expect(buttonNode.text()).toEqual('a:1');
    expect(appBNode.text()).toEqual('a:1');
  })

  it('use-store should work when state changed with promise action', async () => {
    const models = [{
      name: 'modelA',
      state: { a: 0, b: 0 },
      actions: {
        async addA(value: any, state: State) {
          state.a = await Promise.resolve(value);
        }
      },
    }];

    const App = () => {
      const [a, dispatch] = useStore('modelA', s => s.a);
      const handleClick = (e: any) => {
        dispatch("addA", e.target.value)
      }
      return <button onClick={handleClick}>a:{a}</button>;
    };
    const AppB = () => {
      const [a] = useStore('modelA', s => s.a);
      return <div >a:{a}</div>;
    };
    const Context = createSharedStateContext()
    const Test = () => <Context.Provider models={models}><App /><AppB /></Context.Provider>;
    const wrapper = mount(<Test />);
    const buttonNode = wrapper.find("button").at(0);
    const appBNode = wrapper.find("AppB").find("div").at(0);
    expect(buttonNode.text()).toEqual('a:0');
    expect(appBNode.text()).toEqual('a:0');
    await act(async () => {
      buttonNode.simulate("click", { target: { value: 1 } })
      await (global as any).flushPromise(30);
    });
    wrapper.update();
    jest.runAllImmediates()
    expect(buttonNode.text()).toEqual('a:1');
    expect(appBNode.text()).toEqual('a:1');

    await act(async () => {
      buttonNode.simulate("click", { target: { value: 1 } })
      await (global as any).flushPromise(30);
    });
    wrapper.update();
    jest.runAllImmediates()
    expect(buttonNode.text()).toEqual('a:1');
    expect(appBNode.text()).toEqual('a:1');
  })

  it('use-store should work when state changed with promise action usage useStatus', async () => {
    const models = [{
      name: 'modelA',
      state: { a: 0, b: 0 },
      actions: {
        async addA(value: any, state: State) {
          state.a = await Promise.resolve(value);
        },
        async reduceA(value: any, state: State) {
          state.a = await Promise.reject(new Error("err"));
        }
      },
    }];

    const App = () => {
      const [a, dispatch] = useStore('modelA', s => s.a);
      const [status] = useStatus("modelA/reduceA");
      const [statusA] = useStatus("modelA/addA");
      const handleClick = (e: any) => {
        dispatch("addA", e.target.value)
      }
      const handleReduceClick = (e: any) => {
        dispatch("reduceA", e.target.value)
      }
      return <div>
        <button onClick={handleClick}>a:{a}</button>
        <button onClick={handleReduceClick}>a:{status?.error?.toString()}</button>
      </div>;
    };
    const AppB = () => {
      const [a] = useStore('modelA', s => s.a);
      return <div >a:{a}</div>;
    };
    const Context = createSharedStateContext()
    const Test = () => <Context.Provider models={models}><App /><AppB /></Context.Provider>;
    const wrapper = mount(<Test />);
    const buttonNode = wrapper.find("button").at(0);
    const buttonReduceNode = wrapper.find("button").at(1);
    const appBNode = wrapper.find("AppB").find("div").at(0);
    expect(buttonNode.text()).toEqual('a:0');
    expect(appBNode.text()).toEqual('a:0');
    await act(async () => {
      buttonNode.simulate("click", { target: { value: 1 } })
      await (global as any).flushPromise(30);
    });
    wrapper.update();
    jest.runAllImmediates()
    expect(buttonNode.text()).toEqual('a:1');
    expect(appBNode.text()).toEqual('a:1');

    await act(async () => {
      buttonReduceNode.simulate("click", { target: { value: 0 } })
      await (global as any).flushPromise(30);
    });
    wrapper.update();
    jest.runAllImmediates()
    expect(buttonNode.text()).toEqual('a:1');
    expect(buttonReduceNode.text()).toEqual('a:Error: err');
    expect(appBNode.text()).toEqual('a:1');

    wrapper.unmount();
    expect(getContext("modelA")).toBe(null)
  })
}) 
