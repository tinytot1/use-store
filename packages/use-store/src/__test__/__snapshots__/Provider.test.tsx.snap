// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic usage use-store should work 1`] = `
<Test>
  <Memo()
    models={
      Array [
        Object {
          "name": "modelA",
          "state": Object {
            "a": 0,
            "b": 0,
          },
        },
        Object {
          "name": "@STATUS",
          "state": Object {},
        },
      ]
    }
  >
    <Memo()
      model={
        Object {
          "name": "@STATUS",
          "state": Object {},
        }
      }
    >
       
      <Memo()
        model={
          Object {
            "name": "modelA",
            "state": Object {
              "a": 0,
              "b": 0,
            },
          }
        }
      >
         
        <App>
          <button>
            a:
            0
          </button>
        </App>
      </Memo()>
    </Memo()>
  </Memo()>
</Test>
`;
